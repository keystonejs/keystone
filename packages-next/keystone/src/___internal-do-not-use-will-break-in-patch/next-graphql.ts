import { KeystoneConfig } from '@keystone-next/types';
import { initConfig } from '../lib/config/initConfig';
import { createSystem } from '../lib/createSystem';
import { createApolloServerMicro } from '../lib/server/createApolloServer';

export function nextGraphQLAPIRoute(keystoneConfig: KeystoneConfig, prismaClient: any) {
  const initializedKeystoneConfig = initConfig(keystoneConfig);
  const { graphQLSchema, getKeystone } = createSystem(initializedKeystoneConfig);

  const keystone = getKeystone(prismaClient);

  keystone.connect();

  const apolloServer = createApolloServerMicro({
    graphQLSchema,
    createContext: keystone.createContext,
    sessionStrategy: initializedKeystoneConfig.session,
    apolloConfig: initializedKeystoneConfig.graphql?.apolloConfig,
    connectionPromise: keystone.connect(),
  });

  return apolloServer.createHandler({ path: '/api/graphql' });
}
