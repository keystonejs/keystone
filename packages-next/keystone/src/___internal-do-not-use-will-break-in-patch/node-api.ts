import { KeystoneConfig } from '@keystone-next/types';
import { createSystem } from '../lib/createSystem';
import { initConfig } from '../lib/config/initConfig';

export function createListsAPI(config: KeystoneConfig, prismaClient: any) {
  const { createContext, connect } = createSystem(initConfig(config), prismaClient);
  connect();
  return createContext().sudo().lists;
}
