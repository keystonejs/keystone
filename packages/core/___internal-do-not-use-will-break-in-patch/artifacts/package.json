{
  "main": "dist/keystone-6-core-___internal-do-not-use-will-break-in-patch-artifacts.cjs.js",
  "module": "dist/keystone-6-core-___internal-do-not-use-will-break-in-patch-artifacts.esm.js"
}
