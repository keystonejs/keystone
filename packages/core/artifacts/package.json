{
  "main": "dist/keystone-6-core-artifacts.cjs.js",
  "module": "dist/keystone-6-core-artifacts.esm.js"
}
