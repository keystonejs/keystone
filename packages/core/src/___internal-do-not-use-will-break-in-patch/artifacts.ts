export {
  validateCommittedArtifacts,
  generateNodeModulesArtifacts,
  generateCommittedArtifacts,
  getCommittedArtifacts,
} from '../artifacts';
export type { PrismaModule } from '../artifacts';
