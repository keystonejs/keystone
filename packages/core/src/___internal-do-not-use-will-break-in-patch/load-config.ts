export { loadConfigOnce as loadConfig } from '../lib/config/loadConfig';
