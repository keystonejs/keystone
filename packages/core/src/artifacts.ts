import path from 'path';
import { createRequire } from 'module';
import { printSchema, GraphQLSchema } from 'graphql';
import * as fs from 'fs-extra';
import { getGenerator, formatSchema } from '@prisma/sdk';
import { format } from 'prettier';
import type { KeystoneConfig } from './types';
import { confirmPrompt, shouldPrompt } from './lib/prompts';
import { printGeneratedTypes } from './lib/schema-type-printer';
import { ExitError } from './scripts/utils';
import { initialiseLists } from './lib/core/types-for-lists';
import { printPrismaSchema } from './lib/core/prisma-schema';

export function getSchemaPaths(cwd: string) {
  return {
    prisma: path.join(cwd, 'schema.prisma'),
    graphql: path.join(cwd, 'schema.graphql'),
  };
}

type CommittedArtifacts = {
  graphql: string;
  prisma: string;
};

export function getFormattedGraphQLSchema(schema: string) {
  return format(
    '# This file is automatically generated by Keystone, do not modify it manually.\n' +
      '# Modify your Keystone config when you want to change this.\n\n' +
      schema,
    { parser: 'graphql' }
  );
}

export async function getCommittedArtifacts(
  graphQLSchema: GraphQLSchema,
  config: KeystoneConfig
): Promise<CommittedArtifacts> {
  const lists = initialiseLists(config);
  const prismaSchema = printPrismaSchema(
    lists,
    config.db.provider,
    config.db.prismaPreviewFeatures
  );
  return {
    graphql: getFormattedGraphQLSchema(printSchema(graphQLSchema)),
    prisma: await formatPrismaSchema(prismaSchema),
  };
}

let hasEnsuredBinariesExist = false;
async function ensurePrismaBinariesExist() {
  // ensureBinariesExist does a bunch of slightly expensive things
  // so if we can avoid running it a bunch in tests, that's ideal
  if (hasEnsuredBinariesExist) return;
  // we're resolving @prisma/engines from @prisma/sdk
  // because we don't want to depend on @prisma/engines
  // since its version includes a commit hash from https://github.com/prisma/prisma-engines
  // and we just want to use whatever version @prisma/sdk is using
  // also note we use an exact version of @prisma/sdk
  // so if @prisma/sdk suddenly stops depending on @prisma/engines
  // that won't break a released version of Keystone
  // also, we're not just directly importing @prisma/engines
  // since stricter package managers(e.g. pnpm, Yarn Berry)
  // don't allow importing packages that aren't explicitly depended on
  const requireFromPrismaSdk = createRequire(require.resolve('@prisma/sdk'));
  const prismaEngines = requireFromPrismaSdk('@prisma/engines') as typeof import('@prisma/engines');
  await prismaEngines.ensureBinariesExist();
  hasEnsuredBinariesExist = true;
}

async function formatPrismaSchema(schema: string) {
  await ensurePrismaBinariesExist();
  return formatSchema({ schema });
}

async function readFileButReturnNothingIfDoesNotExist(filename: string) {
  try {
    return await fs.readFile(filename, 'utf8');
  } catch (err: any) {
    if (err.code === 'ENOENT') {
      return;
    }
    throw err;
  }
}

export async function validateCommittedArtifacts(
  graphQLSchema: GraphQLSchema,
  config: KeystoneConfig,
  cwd: string
) {
  const artifacts = await getCommittedArtifacts(graphQLSchema, config);
  const schemaPaths = getSchemaPaths(cwd);
  const [writtenGraphQLSchema, writtenPrismaSchema] = await Promise.all([
    readFileButReturnNothingIfDoesNotExist(schemaPaths.graphql),
    readFileButReturnNothingIfDoesNotExist(schemaPaths.prisma),
  ]);
  const outOfDateSchemas = (() => {
    if (writtenGraphQLSchema !== artifacts.graphql && writtenPrismaSchema !== artifacts.prisma) {
      return 'both';
    }
    if (writtenGraphQLSchema !== artifacts.graphql) {
      return 'graphql';
    }
    if (writtenPrismaSchema !== artifacts.prisma) {
      return 'prisma';
    }
  })();
  if (outOfDateSchemas) {
    const message = {
      both: 'Your Prisma and GraphQL schemas are not up to date',
      graphql: 'Your GraphQL schema is not up to date',
      prisma: 'Your Prisma schema is not up to date',
    }[outOfDateSchemas];
    console.log(message);
    const term = {
      both: 'Prisma and GraphQL schemas',
      prisma: 'Prisma schema',
      graphql: 'GraphQL schema',
    }[outOfDateSchemas];
    if (shouldPrompt && (await confirmPrompt(`Would you like to update your ${term}?`))) {
      await writeCommittedArtifacts(artifacts, cwd);
    } else {
      console.log(`Please run keystone postinstall --fix to update your ${term}`);
      throw new ExitError(1);
    }
  }
}

export async function writeCommittedArtifacts(artifacts: CommittedArtifacts, cwd: string) {
  const schemaPaths = getSchemaPaths(cwd);
  await Promise.all([
    fs.writeFile(schemaPaths.graphql, artifacts.graphql),
    fs.writeFile(schemaPaths.prisma, artifacts.prisma),
  ]);
}

export async function generateCommittedArtifacts(
  graphQLSchema: GraphQLSchema,
  config: KeystoneConfig,
  cwd: string
) {
  const artifacts = await getCommittedArtifacts(graphQLSchema, config);
  await writeCommittedArtifacts(artifacts, cwd);
  return artifacts;
}

const nodeAPIJS = (
  cwd: string,
  config: KeystoneConfig
) => `import keystoneConfig from '../../keystone';
import { PrismaClient } from '.prisma/client';
import { createQueryAPI } from '@keystone-6/core/___internal-do-not-use-will-break-in-patch/node-api';
${makeVercelIncludeTheSQLiteDB(cwd, path.join(cwd, 'node_modules/.keystone/next'), config)}

export const query = createQueryAPI(keystoneConfig, PrismaClient);
`;

const nodeAPIDTS = `import { KeystoneListsAPI } from '@keystone-6/core/types';
import { Context } from './types';

export const query: Context['query'];`;

const makeVercelIncludeTheSQLiteDB = (
  cwd: string,
  directoryOfFileToBeWritten: string,
  config: KeystoneConfig
) => {
  if (config.db.provider === 'sqlite') {
    const sqliteDbAbsolutePath = path.resolve(cwd, config.db.url.replace('file:', ''));

    return `import path from 'path';

    path.join(__dirname, ${JSON.stringify(
      path.relative(directoryOfFileToBeWritten, sqliteDbAbsolutePath)
    )});
    path.join(process.cwd(), ${JSON.stringify(path.relative(cwd, sqliteDbAbsolutePath))});
    `;
  }
  return '';
};

const nextGraphQLAPIJS = (
  cwd: string,
  config: KeystoneConfig
) => `import keystoneConfig from '../../../keystone';
import { PrismaClient } from '.prisma/client';
import { nextGraphQLAPIRoute } from '@keystone-6/core/___internal-do-not-use-will-break-in-patch/next-graphql';
${makeVercelIncludeTheSQLiteDB(cwd, path.join(cwd, 'node_modules/.keystone/next'), config)}

export const config = {
  api: {
    bodyParser: false,
  },
};

export default nextGraphQLAPIRoute(keystoneConfig, PrismaClient);
`;

// note the export default config is just a lazy way of going "this is also any"
const nextGraphQLAPIDTS = `export const config: any;
export default config;
`;

export async function generateNodeModulesArtifactsWithoutPrismaClient(
  graphQLSchema: GraphQLSchema,
  config: KeystoneConfig,
  cwd: string
) {
  const lists = initialiseLists(config);

  const dotKeystoneDir = path.join(cwd, 'node_modules/.keystone');
  await Promise.all([
    fs.outputFile(
      path.join(dotKeystoneDir, 'types.d.ts'),
      printGeneratedTypes(graphQLSchema, lists)
    ),
    fs.outputFile(path.join(dotKeystoneDir, 'types.js'), ''),
    ...(config.experimental?.generateNodeAPI
      ? [
          fs.outputFile(path.join(dotKeystoneDir, 'api.js'), nodeAPIJS(cwd, config)),
          fs.outputFile(path.join(dotKeystoneDir, 'api.d.ts'), nodeAPIDTS),
        ]
      : []),
    ...(config.experimental?.generateNextGraphqlAPI
      ? [
          fs.outputFile(
            path.join(dotKeystoneDir, 'next/graphql-api.js'),
            nextGraphQLAPIJS(cwd, config)
          ),
          fs.outputFile(path.join(dotKeystoneDir, 'next/graphql-api.d.ts'), nextGraphQLAPIDTS),
        ]
      : []),
  ]);
}

export async function generateNodeModulesArtifacts(
  graphQLSchema: GraphQLSchema,
  config: KeystoneConfig,
  cwd: string
) {
  await Promise.all([
    generatePrismaClient(cwd),
    generateNodeModulesArtifactsWithoutPrismaClient(graphQLSchema, config, cwd),
  ]);
}

async function generatePrismaClient(cwd: string) {
  const generator = await getGenerator({ schemaPath: getSchemaPaths(cwd).prisma });
  try {
    await generator.generate();
  } finally {
    let closePromise = new Promise<void>(resolve => {
      const child = (generator as any).generatorProcess
        .child as import('child_process').ChildProcess;
      child.once('exit', () => {
        resolve();
      });
    });
    generator.stop();
    await closePromise;
  }
}

export function requirePrismaClient(cwd: string) {
  return require(path.join(cwd, 'node_modules/.prisma/client')).PrismaClient;
}
