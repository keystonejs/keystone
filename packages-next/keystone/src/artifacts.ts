import path from 'path';
import { printSchema, GraphQLSchema } from 'graphql';
import * as fs from 'fs-extra';
import type { KeystoneConfig } from '@keystone-next/types';
import { getGenerator, formatSchema } from '@prisma/sdk';
import { format } from 'prettier';
import { confirmPrompt, shouldPrompt } from './lib/prompts';
import { printGeneratedTypes } from './lib/schema-type-printer';
import { ExitError } from './scripts/utils';
import { initialiseLists } from './lib/core/types-for-lists';
import { printPrismaSchema } from './lib/core/prisma-schema';
import { getDBProvider } from './lib/createSystem';

export function getSchemaPaths(cwd: string) {
  return {
    prisma: path.join(cwd, 'schema.prisma'),
    graphql: path.join(cwd, 'schema.graphql'),
  };
}

type CommittedArtifacts = {
  graphql: string;
  prisma: string;
};

export async function getCommittedArtifacts(
  graphQLSchema: GraphQLSchema,
  config: KeystoneConfig
): Promise<CommittedArtifacts> {
  const lists = initialiseLists(config.lists, getDBProvider(config.db));
  const prismaSchema = printPrismaSchema(
    lists,
    getDBProvider(config.db),
    'node_modules/.prisma/client'
  );
  return {
    graphql: format(printSchema(graphQLSchema), { parser: 'graphql' }),
    prisma: await formatSchema({
      schema: prismaSchema,
    }),
  };
}

async function readFileButReturnNothingIfDoesNotExist(filename: string) {
  try {
    return await fs.readFile(filename, 'utf8');
  } catch (err) {
    if (err.code === 'ENOENT') {
      return;
    }
    throw err;
  }
}

export async function validateCommittedArtifacts(
  graphQLSchema: GraphQLSchema,
  config: KeystoneConfig,
  cwd: string
) {
  const artifacts = await getCommittedArtifacts(graphQLSchema, config);
  const schemaPaths = getSchemaPaths(cwd);
  const [writtenGraphQLSchema, writtenPrismaSchema] = await Promise.all([
    readFileButReturnNothingIfDoesNotExist(schemaPaths.graphql),
    readFileButReturnNothingIfDoesNotExist(schemaPaths.prisma),
  ]);
  const outOfDateSchemas = (() => {
    if (writtenGraphQLSchema !== artifacts.graphql && writtenPrismaSchema !== artifacts.prisma) {
      return 'both';
    }
    if (writtenGraphQLSchema !== artifacts.graphql) {
      return 'graphql';
    }
    if (writtenPrismaSchema !== artifacts.prisma) {
      return 'prisma';
    }
  })();
  if (outOfDateSchemas) {
    const message = {
      both: 'Your Prisma and GraphQL schemas are not up to date',
      graphql: 'Your GraphQL schema is not up to date',
      prisma: 'Your GraphQL schema is not up to date',
    }[outOfDateSchemas];
    console.log(message);
    const term = {
      both: 'Prisma and GraphQL schemas',
      prisma: 'Prisma schema',
      graphql: 'GraphQL schema',
    }[outOfDateSchemas];
    if (shouldPrompt && (await confirmPrompt(`Would you like to update your ${term}?`))) {
      await writeCommittedArtifacts(artifacts, cwd);
    } else {
      console.log(`Please run keystone-next postinstall --fix to update your ${term}`);
      throw new ExitError(1);
    }
  }
}

export async function writeCommittedArtifacts(artifacts: CommittedArtifacts, cwd: string) {
  const schemaPaths = getSchemaPaths(cwd);
  await Promise.all([
    fs.writeFile(schemaPaths.graphql, artifacts.graphql),
    fs.writeFile(schemaPaths.prisma, artifacts.prisma),
  ]);
}

export async function generateCommittedArtifacts(
  graphQLSchema: GraphQLSchema,
  config: KeystoneConfig,
  cwd: string
) {
  const artifacts = await getCommittedArtifacts(graphQLSchema, config);
  await writeCommittedArtifacts(artifacts, cwd);
  return artifacts;
}

const nodeAPIJS = (
  cwd: string,
  config: KeystoneConfig
) => `import keystoneConfig from '../../keystone';
import { PrismaClient } from '.prisma/client';
import { createListsAPI } from '@keystone-next/keystone/___internal-do-not-use-will-break-in-patch/node-api';
${makeVercelIncludeTheSQLiteDB(cwd, path.join(cwd, 'node_modules/.keystone/next'), config)}

export const lists = createListsAPI(keystoneConfig, PrismaClient);
`;

const nodeAPIDTS = `import { KeystoneListsAPI } from '@keystone-next/types';
import { KeystoneListsTypeInfo } from './types';

export const lists: KeystoneListsAPI<KeystoneListsTypeInfo>;`;

const makeVercelIncludeTheSQLiteDB = (
  cwd: string,
  directoryOfFileToBeWritten: string,
  config: KeystoneConfig
) => {
  if (config.db.adapter === 'prisma_sqlite' || config.db.provider === 'sqlite') {
    const sqliteDbAbsolutePath = path.resolve(cwd, config.db.url.replace('file:', ''));

    return `import path from 'path';

    path.join(__dirname, ${JSON.stringify(
      path.relative(directoryOfFileToBeWritten, sqliteDbAbsolutePath)
    )});
    path.join(process.cwd(), ${JSON.stringify(path.relative(cwd, sqliteDbAbsolutePath))});
    `;
  }
  return '';
};

const nextGraphQLAPIJS = (
  cwd: string,
  config: KeystoneConfig
) => `import keystoneConfig from '../../../keystone';
import { PrismaClient } from '.prisma/client';
import { nextGraphQLAPIRoute } from '@keystone-next/keystone/___internal-do-not-use-will-break-in-patch/next-graphql';
${makeVercelIncludeTheSQLiteDB(cwd, path.join(cwd, 'node_modules/.keystone/next'), config)}

export const config = {
  api: {
    bodyParser: false,
  },
};

export default nextGraphQLAPIRoute(keystoneConfig, PrismaClient);
`;

// note the export default config is just a lazy way of going "this is also any"
const nextGraphQLAPIDTS = `export const config: any;
export default config;
`;

export async function generateNodeModulesArtifacts(
  graphQLSchema: GraphQLSchema,
  config: KeystoneConfig,
  cwd: string
) {
  const lists = initialiseLists(config.lists, getDBProvider(config.db));

  const printedSchema = printSchema(graphQLSchema);
  const dotKeystoneDir = path.join(cwd, 'node_modules/.keystone');
  await Promise.all([
    generatePrismaClient(cwd),
    fs.outputFile(
      path.join(dotKeystoneDir, 'types.d.ts'),
      printGeneratedTypes(printedSchema, graphQLSchema, lists)
    ),
    fs.outputFile(path.join(dotKeystoneDir, 'types.js'), ''),
    ...(config.experimental?.generateNodeAPI
      ? [
          fs.outputFile(path.join(dotKeystoneDir, 'api.js'), nodeAPIJS(cwd, config)),
          fs.outputFile(path.join(dotKeystoneDir, 'api.d.ts'), nodeAPIDTS),
        ]
      : []),
    ...(config.experimental?.generateNextGraphqlAPI
      ? [
          fs.outputFile(
            path.join(dotKeystoneDir, 'next/graphql-api.js'),
            nextGraphQLAPIJS(cwd, config)
          ),
          fs.outputFile(path.join(dotKeystoneDir, 'next/graphql-api.d.ts'), nextGraphQLAPIDTS),
        ]
      : []),
  ]);
}

async function generatePrismaClient(cwd: string) {
  const generator = await getGenerator({ schemaPath: getSchemaPaths(cwd).prisma });
  await generator.generate();
  generator.stop();
}

export function requirePrismaClient(cwd: string) {
  return require(path.join(cwd, 'node_modules/.prisma/client')).PrismaClient;
}
