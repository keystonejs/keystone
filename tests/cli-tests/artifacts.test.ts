import { ExitError } from '@keystone-6/core/___internal-do-not-use-will-break-in-patch/artifacts'

import {
  basicKeystoneConfig,
  getFiles,
  recordConsole,
  cliMock,
  schemas,
  symlinkKeystoneDeps,
  testdir,
} from './utils'

describe.each(['postinstall', ['build', '--frozen']])('%s', command => {
  test('logs an error and exits with 1 when the schemas do not exist', async () => {
    const cwd = await testdir({
      ...symlinkKeystoneDeps,
      'keystone.js': basicKeystoneConfig,
    })

    const recording = recordConsole()
    await expect(cliMock(cwd, command)).rejects.toEqual(new ExitError(1))

    expect(recording()).toMatchInlineSnapshot(`"Your Prisma and GraphQL schemas are not up to date"`)
  })
})

describe('prisma migrate status', () => {
  test('logs an error and exits with 1 when the schemas do not exist', async () => {
    const cwd = await testdir({
      ...symlinkKeystoneDeps,
      'keystone.js': basicKeystoneConfig,
    })
    await expect(cliMock(cwd, ['build', '--no-ui', '--frozen'])).rejects.toEqual(new ExitError(1))

    const recording = recordConsole()
    await expect(cliMock(cwd, ['prisma', '--frozen', 'migrate', 'status'])).rejects.toEqual(new ExitError(1))

    expect(recording()).toMatchInlineSnapshot(`"Your Prisma and GraphQL schemas are not up to date"`)
  })
})

const schemasMatch = ['schema.prisma', 'schema.graphql']

// a lot of these cases are also the same for prisma and build commands but we don't include them here
// because when they're slow and then run the same code as the postinstall command
// (and in the case of the build command we need to spawn a child process which would make each case take a _very_ long time)
describe('postinstall', () => {
  test('updates the schemas without prompting when --fix is passed', async () => {
    const cwd = await testdir({
      ...symlinkKeystoneDeps,
      'keystone.js': basicKeystoneConfig,
    })

    const recording = recordConsole()
    await cliMock(cwd, ['postinstall', '--fix'])
    const files = await getFiles(cwd, schemasMatch)

    expect(files).toEqual(await getFiles(`${__dirname}/fixtures/basic-project`, schemasMatch))
    expect(recording()).toMatchInlineSnapshot(`"? Generated GraphQL and Prisma schemas"`)
  })

  test("does not prompt, error or modify the schemas if they're already up to date", async () => {
    const cwd = await testdir({
      ...symlinkKeystoneDeps,
      ...schemas,
      'keystone.js': basicKeystoneConfig,
    })

    const recording = recordConsole()
    await cliMock(cwd, 'postinstall')
    const files = await getFiles(cwd, schemasMatch)

    expect(files).toEqual(await getFiles(`${__dirname}/fixtures/basic-project`, schemasMatch))
    expect(recording()).toMatchInlineSnapshot(`"? GraphQL and Prisma schemas are up to date"`)
  })

  test('writes the correct node_modules files', async () => {
    const cwd = await testdir({
      ...symlinkKeystoneDeps,
      ...schemas,
      'keystone.js': basicKeystoneConfig,
    })

    const recording = recordConsole()
    await cliMock(cwd, 'postinstall')

    expect(await getFiles(cwd, ['node_modules/.keystone/**/*'])).toMatchSnapshot()
    expect(recording()).toMatchInlineSnapshot(`"? GraphQL and Prisma schemas are up to date"`)
  })
})
