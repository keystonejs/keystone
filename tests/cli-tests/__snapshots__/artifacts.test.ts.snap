// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postinstall writes the correct node_modules files 1`] = `
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ node_modules/.keystone/types.ts ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
/* eslint-disable */

export type TodoWhereUniqueInput = {
  readonly id?: string | null
}

export type TodoWhereInput = {
  readonly AND?: ReadonlyArray<TodoWhereInput> | TodoWhereInput | null
  readonly OR?: ReadonlyArray<TodoWhereInput> | TodoWhereInput | null
  readonly NOT?: ReadonlyArray<TodoWhereInput> | TodoWhereInput | null
  readonly id?: IDFilter | null
  readonly title?: StringFilter | null
}

export type IDFilter = {
  readonly equals?: string | null
  readonly in?: ReadonlyArray<string> | string | null
  readonly notIn?: ReadonlyArray<string> | string | null
  readonly lt?: string | null
  readonly lte?: string | null
  readonly gt?: string | null
  readonly gte?: string | null
  readonly not?: IDFilter | null
}

export type StringFilter = {
  readonly equals?: string | null
  readonly in?: ReadonlyArray<string> | string | null
  readonly notIn?: ReadonlyArray<string> | string | null
  readonly lt?: string | null
  readonly lte?: string | null
  readonly gt?: string | null
  readonly gte?: string | null
  readonly contains?: string | null
  readonly startsWith?: string | null
  readonly endsWith?: string | null
  readonly not?: NestedStringFilter | null
}

export type NestedStringFilter = {
  readonly equals?: string | null
  readonly in?: ReadonlyArray<string> | string | null
  readonly notIn?: ReadonlyArray<string> | string | null
  readonly lt?: string | null
  readonly lte?: string | null
  readonly gt?: string | null
  readonly gte?: string | null
  readonly contains?: string | null
  readonly startsWith?: string | null
  readonly endsWith?: string | null
  readonly not?: NestedStringFilter | null
}

export type TodoOrderByInput = {
  readonly id?: OrderDirection | null
  readonly title?: OrderDirection | null
}

export type OrderDirection =
  | 'asc'
  | 'desc'

export type TodoUpdateInput = {
  readonly title?: string | null
}

export type TodoUpdateArgs = {
  readonly where: TodoWhereUniqueInput
  readonly data: TodoUpdateInput
}

export type TodoCreateInput = {
  readonly title?: string | null
}

export type KeystoneAdminUIFieldMetaIsNonNull =
  | 'read'
  | 'create'
  | 'update'

export type KeystoneAdminUIFieldMetaCreateViewFieldMode =
  | 'edit'
  | 'hidden'

export type KeystoneAdminUIFieldMetaListViewFieldMode =
  | 'read'
  | 'hidden'

export type KeystoneAdminUIFieldMetaItemViewFieldMode =
  | 'edit'
  | 'read'
  | 'hidden'

export type KeystoneAdminUIFieldMetaItemViewFieldPosition =
  | 'form'
  | 'sidebar'

export type QueryMode =
  | 'default'
  | 'insensitive'

export type KeystoneAdminUISortDirection =
  | 'ASC'
  | 'DESC'

type ResolvedTodoCreateInput = {
  id?: import('@prisma/client').Prisma.TodoCreateInput['id']
  title?: import('@prisma/client').Prisma.TodoCreateInput['title']
}
type ResolvedTodoUpdateInput = {
  id?: undefined
  title?: import('@prisma/client').Prisma.TodoUpdateInput['title']
}

export declare namespace Lists {
  export type Todo<Session = any> = import('@keystone-6/core').ListConfig<Lists.Todo.TypeInfo<Session>>
  namespace Todo {
    export type Item = import('@prisma/client').Todo
    export type TypeInfo<Session = any> = {
      key: 'Todo'
      isSingleton: false
      fields: 'id' | 'title'
      item: Item
      inputs: {
        where: TodoWhereInput
        uniqueWhere: TodoWhereUniqueInput
        create: TodoCreateInput
        update: TodoUpdateInput
        orderBy: TodoOrderByInput
      }
      prisma: {
        create: ResolvedTodoCreateInput
        update: ResolvedTodoUpdateInput
      }
      all: __TypeInfo<Session>
    }
  }
}
export type Context<Session = any> = import('@keystone-6/core/types').KeystoneContext<TypeInfo<Session>>
export type Config<Session = any> = import('@keystone-6/core/types').KeystoneConfig<TypeInfo<Session>>

export type TypeInfo<Session = any> = {
  lists: {
    readonly Todo: Lists.Todo.TypeInfo<Session>
  }
  prisma: import('@prisma/client').PrismaClient
  session: Session
}

type __TypeInfo<Session = any> = TypeInfo<Session>

export type Lists<Session = any> = {
  [Key in keyof TypeInfo['lists']]?: import('@keystone-6/core').ListConfig<TypeInfo<Session>['lists'][Key]>
} & Record<string, import('@keystone-6/core').ListConfig<any>>

export {}

`;
